### PR TITLE
Permission management

### LINKED ISSUE #46: Implement Terrafrom resource azuredevops_git_permissions
---

Migrated from https://github.com/microsoft/terraform-provider-azuredevops/issues/153
Originally created by @tmeckel on *Fri, 25 Oct 2019 12:17:24 GMT*

---

## Community Note

* Please vote on this issue by adding a 👍
  [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)
  to the original issue to help the community and maintainers prioritize this
  request
* Please do not leave "+1" or "me too" comments, they generate extra noise for
  issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull
  request, please leave a comment

## Description

As a developer, I want the ability to manage the permissions of all Git
repositories inside a project, a single Git Repository and a single branch
inside a Git repository in Azure DevOps. Thus a new Terraform resource shall be
provided.

The resource shall provide three attributes to define the scope on which the
permissions should be applied:

* `project_id` (required)
* `repository_id` (optional)
* `branch_id` (optional)

Logic:

1. to apply permissions to all Git repositories inside a project set `project_id`
2. to apply permissions to a specific Git repository inside a project set `project_id` and `repository_id`
3. to apply permissions to a branch inside a Git repository of a project set `project_id`, `repository_id` and `branch_id`

The resource shall issue an error if:

1. the `project_id` is not set
2. the `branch_id` is set without having set `repository_id`

### New or Affected Resource(s)

* azuredevops_git_permissions

### Potential Terraform Configuration

```hcl

data "azuredevops_project" "project" {
   # Exports: id
   project_name = "Test Project"
}

resource "azuredevops_group" "project-group" {
    # Exports: id, descriptor (Security Descriptor)
    project_id = data.azuredevops_project.project.id
    name       = "Test Group"
}

data "azuredevops_git_repository" "repo" {
    name = "RepoV2"
}

data "azuredevops_git_branch" "branch_foobar" {
    name          = "Foobar"
    repository_id = data.azuredevops_git_repository.repo.id
    # Id            = /refs/heads/6d0061007300740065007200
}

resource "azuredevops_git_permissions" "perm" {
    # Security Namespace: 2e9eb7ed-3c0a-47d4-87c1-0ffdd275fd87

    // Build ACL token from project id (GUID) as follows
    // repoV2/<projectGUID>/<repoGUID>/<branchGUID>
    project_id    = data.azuredevops_project.project.id
    repository_id = data.azuredevops_git_repository.repo.id
    branch_id     = data.azuredevops_git_branch.branch_foobar.id

    principals = [
        data.azuredevops_project_user.lookup-project-user.id_descriptor
        azuredevops_user.project-user.id_descriptor
        # id_descriptor returns the decoded descriptor value (identity descriptor)
        # id => aad.YmZhYmE4ZDUtZWMzZC03ZjIzLWE4NDktMDM3ZjEwMTEwNDZh
        # id_descriptor => Microsoft.IdentityModel.Claims.ClaimsIdentity;2eb07b91-f9f8-437b-a89e-59f55e565c11\\EE03B31@UNION-INVESTMENT.DE
        azuredevops_group.project-group.id_descriptor
    ]

    permissions = {
        "Administer"   = "Allow"
        "ForcePush"    = "Deny"
        "CreateBranch" = "NotSet"
    }

}

```

### Acceptance Criteria

[] Include Unit tests for the new operation 

[] Acceptance (integration) test to validate repo creation/modification, 
   following the patterns from the azuredevops_build_definition resource

### References

* #0000

### LINKED ISSUE #47: Implement Terrafrom resource azuredevops_project_permissions
---

Migrated from https://github.com/microsoft/terraform-provider-azuredevops/issues/104
Originally created by @tmeckel on *Mon, 14 Oct 2019 15:31:03 GMT*

---

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

As a developer, I want the ability to manage the permissions of a project in
Azure DevOps. Thus a new Terraform resource shall be provided.

### New or Affected Resource(s)

* azuredevops_project_permissions

### Potential Terraform Configuration

```hcl

data "azuredevops_project" "project" {
   # Exports: id
   project_name = "Test Project"
}

data "azuredevops_user" "lookup-user" {
   principalname_name  = "TestUser@contoso.com"
}

resource "azuredevops_group" "project-group" {
    project_id = azuredevops_project.project.id
    name       = "Test Group"
}

resource "azuredevops_project_permissions" "perm-project-group" {
    # Security Namespace: 52d39943-cb85-4d7f-8fa8-c6baac873819

    # Build ACL token from project id (GUID) as follows
    # $PROJECT:vstfs:///Classification/TeamProject/<ProjectGUID>
    project_id = data.azuredevops_project.project.id

    principals = [
        data.azuredevops_user.lookup-user.id_descriptor
        azuredevops_group.project-group.id_descriptor
    ]

    permissions = {
        "Enumerate"            = "Allow"
        "Delete"               = "Deny"
        "Publish_test_results" = "NotSet"
    }
}

```
### Acceptance Criteria

[] Include Unit tests for the new operation 

[] Acceptance (integration) test to validate repo creation/modification, 
   following the patterns from the azuredevops_build_definition resource

### References

* #127
* #103
* #84 
* [Security REST Api](https://docs.microsoft.com/en-us/rest/api/azure/devops/security/?view=azure-devops-rest-5.1)
* https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/security/client.go
* [Query Security Namespaces and their definitions](https://docs.microsoft.com/en-us/rest/api/azure/devops/security/security%20namespaces/query?view=azure-devops-rest-5.1)

### PR DESCRIPTION
**Pull request by tmeckel**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Initially filed as Pull Request at https://github.com/microsoft/terraform-provider-azuredevops/pull/238

Issue Number:

* Closes https://github.com/microsoft/terraform-provider-azuredevops/issues/104
* Closes https://github.com/microsoft/terraform-provider-azuredevops/issues/153
* Closes #47 
* Closes #46 

The current implementation of the provider does not support the management of permission for project and git repositories.

This PR provides an implementation for two new resources (`azuredevops_project_permissions`, `azuredevops_git_permissions`) as described in the issues mentioned above and provides a framework for an easy implementation for other permission related Terraform resources (Security Namespace). 

In addition the single-value data source `azuredevops_git_repository` has been added.

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

* [x] Yes
* [ ] No

* Upgraded terraform-plugin-sdk to v1.10.0, to have the ability to use [`RequiredWith`](https://github.com/hashicorp/terraform-plugin-sdk/blob/994e06a283d731b223ba5d2c120f5e0e1c9f7d1d/helper/schema/schema.go#L231) validation in resources schema.
* Removed reference to `github.com/microsoft/terraform-provider-azuredevops`

## Does this introduce a breaking change?

* [ ] Yes
* [x] No

## Any relevant logs, error output, etc?

[N/A]

### CODE REVIEW

**Comment by tmeckel** [comment]
@nmiodice, @xuzhang3 I changed the `scripts/unittest.sh` (924eedcc63fda6688ef2d5ca9bc9e409b2180b66) and `scripts/acctest.sh` (35d9884e7d140a61404acc4f24ee9b2e234cefff) scripts so that the build pipeline will not fail on skipped tests.

**Comment by tmeckel** [comment]
@EliiseS @nmiodice @xuzhang3  In addition I fixed some linting errors in `azuredevops/internal/service/taskagent/resource_variable_group.go` (9660e17ac66b33c1800a14a7be053ddc7e521ac5)

File: azuredevops/internal/service/permissions/resource_git_permissions.go
Comment (EliiseS) [inline azuredevops/internal/service/permissions/resource_git_permissions.go]: Super nit-pick: Could you move the functions so that the functions called by the resource are first?	
```golang
                Create: resourceGitPermissionsCreate,
		Read:   resourceGitPermissionsRead,
		Update: resourceGitPermissionsUpdate,
		Delete: resourceGitPermissionsDelete,
```

File: azuredevops/internal/service/permissions/resource_project_permissions.go
Comment (EliiseS) [inline azuredevops/internal/service/permissions/resource_project_permissions.go]: nit: Remove a empty line

File: azuredevops/internal/service/permissions/resource_project_permissions_test.go
Comment (EliiseS) [inline azuredevops/internal/service/permissions/resource_project_permissions_test.go:24]: Super nit: Can you switch the order of these functions, so that the public functions are written first?

File: azuredevops/internal/utils/debug/Debug.go
Comment (EliiseS) [inline azuredevops/internal/utils/debug/Debug.go]: Why is this needed?

**Review by EliiseS — APPROVED** [review]

**Comment by EliiseS** [comment]
It looks great! But the build is broken and I think we should  have a build that runs the test before we can merge this. The build is being blocked by #10

File: azuredevops/internal/service/permissions/resource_project_permissions.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_project_permissions.go]: Not sure which line 😞

File: azuredevops/internal/service/permissions/resource_project_permissions_test.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_project_permissions_test.go:24]: Done

File: azuredevops/internal/utils/debug/Debug.go
Comment (tmeckel) [reply azuredevops/internal/utils/debug/Debug.go]: Personally I got fed up with adding this infamous `time.Sleep` to be able to attach the debugger to a running Go process. So I thought it could be worth adding a function that could be controlled via some environment variables. If we'd use this for every resource or data source, people could attach debuggers even for a official version of the provider, so that they could provide us with more detailed error reports or even pull requests at best.

**Comment by tmeckel** [comment]
> 
> 
> It looks great! But the build is broken and I think we should have a build that runs the test before we can merge this. The build is being blocked by #10

@EliiseS I totally agree, but I've got the impression that we already lost momentum here again. Since the provider has been migrated to Hashicorp I don't see much progress.  It got kinda silent here. Not much from @nmiodice or @xuzhang3 either. The build pipeline has to be fixed QUICKLY! Thought about removing the log messages from the `Skip` function be replacing them with `SkipNow` which hopefully let the build command pass in the pipeline, but I don't wanna change too much code, where the root cause lies somewhere else.

Who is working on #10? Nick? @xuzhang3? You? Someone from Hashicorp? A slightly unsatisfactory situation right now again. 😞

File: azuredevops/internal/service/permissions/resource_project_permissions.go
Comment (EliiseS) [reply azuredevops/internal/service/permissions/resource_project_permissions.go]: Aah, I thought there was two empty lines! Nevermind

File: azuredevops/internal/utils/debug/Debug.go
Comment (EliiseS) [reply azuredevops/internal/utils/debug/Debug.go]: Why do we need time.sleep to attach the debugger? What have I missed?

Comment (tmeckel) [reply azuredevops/internal/utils/debug/Debug.go]: How are you debugging the provider?

Comment (xuzhang3) [reply azuredevops/internal/utils/debug/Debug.go]: > Why do we need time.sleep to attach the debugger? What have I missed?

Mostly for debug purpose.  It 's some kind like bus station waiting for passengers. 
[IDEA debug go application](https://blog.jetbrains.com/go/2019/02/06/debugging-with-goland-getting-started/) roll a little down to `Debugging a running application on the local machine
`

Comment (xuzhang3) [reply azuredevops/internal/utils/debug/Debug.go]: Hi @tmeckel, I prefer acceptance test. As acceptance test will create the real resource, when the acceptance test is executing, I can debug the provider source code. Combine with `TF_LOG=TRACE`, `TF_LOG_PATH=log_file_path` configures. This covers most scenarios for me.
I don't know what develop environment you are. I use Goland, it might different from you.

Comment (tmeckel) [reply azuredevops/internal/utils/debug/Debug.go]: From my point of it's very valuable to have an debugger attached to a running provider because you can analyze current values in variables what gives you often a hint why things go wrong and with wrong I mean, you don't get a crash dump (panic) but simply your code doesn't work as expected.

I think I mostly learned about Terraform and it's providers by stepping through using a debugger. Another advantage using a debugger is the fact that you don't rely on people adding (meaningful) log messages. My experience is that this is mostly not the case 😁 

And about @EliiseS question why the wait (sleep) is needed: Terraform starts/executes the providers as a new process. So to have the ability to attach to the "spawned" provider process you must have some kind of waiting mechanism, so that you have the time to extract the process id of the provider process and use VSCode (dlv) to attach to that process. Go doesn't have this luxury like in .Net or Windows Native applications where you can add an `Debugger.Launch()` which automatically attaches Visual Studio to the running process.

https://github.com/hashicorp/terraform/issues/20466
https://github.com/hashicorp/terraform/pull/16305

Comment (xuzhang3) [reply azuredevops/internal/utils/debug/Debug.go]: Recommend debug with acceptance test during your development. Don't need additional code for debug purpose and we can debug directly from the source code not the binary file.  More smooth and slight.

**Comment by xuzhang3** [comment]
@tmeckel  Can split this PR into small PRs? Dependency upgrade can be included  in separate PR. One PR for one feature or function will be great.

File: azuredevops/internal/utils/debug/Debug.go
Comment (tmeckel) [reply azuredevops/internal/utils/debug/Debug.go]: Okay ... so what's the decision here? Keep the code? Remove it?

Comment (EliiseS) [reply azuredevops/internal/utils/debug/Debug.go]: How are you debugging the provider? I'm unable to get debug to work for with the acceptance tests, but debug for unit tests I can just press the `debug test` button above the test in VSCode.

Comment (tmeckel) [reply azuredevops/internal/utils/debug/Debug.go]: Debugging unit tests won't give you any insights to the interaction between Terraform, the provider and the backend API that the provider is using, because unit tests shouldn't use any APIs aside the unit (module) which is under test. That's why @nmiodice (or someone other clever person) introduced the usage of API mocks for the DevOps provider. 

This is also the reason why @xuzhang3 favors the use of acceptance tests so much. The unit tests don't test all components together. As good and important as acceptance testing is, I believe that certain bugs cannot be solved by using acceptance testing alone, or the solution takes much longer. Therefore I use the GO Debugger to analyze and solve persistent problems and to follow the inner process. 

I always proceed the same way:

1. create a HCL file that calls up the exact area of the provider where the problem is.
2. add a `time.Sleep` method typically at a function that is called by Terraform like `read`, `update` etc.
3. compiling the provider with additional debug flags. That's why I once added those flags to `build.sh` and filed a PR 😁 
4. Call `terraform plan` or `apply`
5. determine the process ID of the provider that was started by Terraform
6. entering the process ID in the `launch.json` of VSCode
7. press F5 at VSCode to attach the debugger to the provider
8. perform debugging

How to debug Go with a debugger is described here: https://github.com/golang/vscode-go/blob/master/docs/debugging.md

**Comment by tmeckel** [comment]
> 
> 
> @tmeckel Can we split this huge PR into small PRs? Dependency upgrade can be included in separate PR. One PR for one feature or function will be great.

I know the pull request is big, but to break it up into small pieces again would take a lot of time for me and I mostly don't get paid for this work, especially for work that's related to get a PR merged, because the customers I'm working with, use an own version of the provider that has those changes incorporated already.

File: azuredevops/internal/utils/debug/Debug.go
Comment (xuzhang3) [reply azuredevops/internal/utils/debug/Debug.go]: > How are you debugging the provider? I'm unable to get debug to work for with the acceptance tests, but debug for unit tests I can just press the `debug test` button above the test in VSCode.

`dlv` is a good tool for debugging [delve](https://github.com/go-delve/delve). 
Usage example (windows environment): 
- open CMD 
- cd to the test file directory
- `dlv test .  -- -test.run=TestAccExampleService_basic -test.v`
- enjoy your debug steps

Comment (xuzhang3) [reply azuredevops/internal/utils/debug/Debug.go]: > Okay ... so what's the decision here? Keep the code? Remove it?

The debug help file should be removed. You can keep this help files in your person branch.   [dlv](https://github.com/go-delve/delve/) may be an optional for you for debug.

Comment (xuzhang3) [reply azuredevops/internal/utils/debug/Debug.go]: [GDB](https://golang.org/doc/gdb) is also an optional for debugging. I will update the development document later when I got time.

Comment (tmeckel) [reply azuredevops/internal/utils/debug/Debug.go]: Removed the code

**Comment by tmeckel** [comment]
HALLELUJAH!!! The pipeline passed 🚀 Well ... without acceptance tests ... Not that great that reviewers currently must execute the acceptance tests locally. 😞 Requesting review @EliiseS , @xuzhang3 @nmiodice 

Oh and I again had to update the vendor directory. Added 50 files to the number of _Files changed_ ... the number is insane already 😁 💥

**Comment by xuzhang3** [comment]
@tmeckel  I will be late for this PR. I'm busy this week.

**Comment by tmeckel** [comment]
@xuzhang3 Thanks for letting me know

File: vendor/golang.org/x/sys/unix/README.md
Comment (seqdan) [inline vendor/golang.org/x/sys/unix/README.md:92]: Wouldn't "may" be correct here?

File: website/docs/d/data_git_repository.html.markdown
Comment (seqdan) [inline website/docs/d/data_git_repository.html.markdown]: Awesome to have documentation here; thanks for adding it! The feature looks promising. - Just a little typo: "information" should not have an "s", as it is uncountable (https://dictionary.cambridge.org/grammar/british-grammar/information).

File: vendor/golang.org/x/sys/unix/README.md
Comment (xuzhang3) [reply vendor/golang.org/x/sys/unix/README.md:92]: Hi @seqdan  The files in vendor  folder  are dependencies, we cannot change this files.  We update this files with go command `go mod vendor` and `go mod tidy`.

Comment (seqdan) [reply vendor/golang.org/x/sys/unix/README.md:92]: Oh, I understand! Sorry! - I'm observing this exciting project, but my skills are not sufficient to contribute. Apologies.

Comment (xuzhang3) [reply vendor/golang.org/x/sys/unix/README.md:92]: Hi @seqdan , No need apologies. It's an opensource project, any questions and contributions are welcome. You can find the [contribute document](https://github.com/terraform-providers/terraform-provider-azuredevops/blob/master/docs/contributing.md) here.

File: azuredevops/internal/acceptancetests/data_git_repository_test.go
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/data_git_repository_test.go:30]: Is this config was just for verify if `terraform apply` works?  No check is configured for this step

File: azuredevops/internal/acceptancetests/resource_group_membership_test.go
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_group_membership_test.go]: This issue link is still in microsoft repo.

File: azuredevops/internal/service/permissions/resource_project_permissions.go
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_project_permissions.go]: Project permissions does not support import operation . This function can be removed.

File: azuredevops/internal/service/permissions/resource_project_permissions.go
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_project_permissions.go]: Project permissions does not support import operation . This code block can be removed.

File: azuredevops/internal/service/permissions/utils/namespaces_test.go
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/utils/namespaces_test.go]: Comment out  code can be removed

File: azuredevops/internal/service/permissions/utils/namespaces.go
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/utils/namespaces.go]: Comment out code can be removed

File: azuredevops/internal/service/permissions/utils/namespaces.go
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/utils/namespaces.go]: Comment out code can be removed

File: azuredevops/internal/service/permissions/resource_project_permissions_test.go
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_project_permissions_test.go]: If we have nothing to init, can we remove this function?

File: website/docs/r/git_permissions.html.markdown
Comment (xuzhang3) [inline website/docs/r/git_permissions.html.markdown:68]: `terraform apply` failed with this example
```
Error: No branch found with name [master] in repository with id [4ca8ea47-f9d7-4ca3-97b0-e1eaad8e41f0]

  on main.tf line 51, in resource "azuredevops_git_permissions" "project-git-branch-permissions":
  51: resource "azuredevops_git_permissions" "project-git-branch-permissions" {

```

File: website/docs/r/project_permissions.html.markdown
Comment (xuzhang3) [inline website/docs/r/project_permissions.html.markdown]: `follwing` should be `following`

File: azuredevops/internal/acceptancetests/data_git_repository_test.go
Comment (tmeckel) [reply azuredevops/internal/acceptancetests/data_git_repository_test.go:30]: I simply copied the code that @nmiodice created to test the multi value data source in `azuredevops/internal/acceptancetests/data_git_repositories_test.go` If there is an issue here we have to check about the acceptance test in this data source as well.

File: azuredevops/internal/acceptancetests/resource_group_membership_test.go
Comment (tmeckel) [reply azuredevops/internal/acceptancetests/resource_group_membership_test.go]: Changed

File: azuredevops/internal/service/permissions/resource_project_permissions.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_project_permissions.go]: Removed

File: azuredevops/internal/service/permissions/resource_project_permissions.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_project_permissions.go]: Removed

File: azuredevops/internal/service/permissions/utils/namespaces_test.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/utils/namespaces_test.go]: Removed

File: azuredevops/internal/service/permissions/utils/namespaces.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/utils/namespaces.go]: Removed

File: azuredevops/internal/service/permissions/utils/namespaces.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/utils/namespaces.go]: Removed

File: azuredevops/internal/service/permissions/resource_project_permissions_test.go
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_project_permissions_test.go]: Removed

File: website/docs/r/project_permissions.html.markdown
Comment (tmeckel) [reply website/docs/r/project_permissions.html.markdown]: Corrected

File: website/docs/r/git_permissions.html.markdown
Comment (tmeckel) [reply website/docs/r/git_permissions.html.markdown:68]: Fixed HCL sample code 

image

File: website/docs/d/data_git_repository.html.markdown
Comment (tmeckel) [reply website/docs/d/data_git_repository.html.markdown]: Corrected

**Comment by xuzhang3** [comment]
Hi @tmeckel   I need some extra test for this huge PR 🍡.   I found bug when I execute the example configuration script in `git_permissions`, I'm need investigate this issue. 
image

#### Steps to reproduce
- execute `terraform apply --auto-approve`
- change some configuration
- execute `terraform apply --auto-approve` again

**Comment by tmeckel** [comment]
> 
> 
> Hi @tmeckel I need some extra test for this huge PR 🍡. I found bug when I execute the example configuration script in `git_permissions`, I'm need investigate this issue.
> image
> #### Steps to reproduce
> 
>     * execute `terraform apply --auto-approve`
> 
>     * change some configuration
> 
>     * execute `terraform apply --auto-approve` again

I can reproduce the error on my side.

**Comment by tmeckel** [comment]
> 
> 
> Hi @tmeckel I need some extra test for this huge PR 🍡. I found bug when I execute the example configuration script in `git_permissions`, I'm need investigate this issue.
> image
> #### Steps to reproduce
> 
>     * execute `terraform apply --auto-approve`
> 
>     * change some configuration
> 
>     * execute `terraform apply --auto-approve` again

@xuzhang3 if you remove the `default_branch` property it all works as expected. Oh you don't have to change anything in the HCL code, the error occurs while refreshing the state. So I suspected `resourceGitRepositoryRead` to be the culprit, but I don't see any reference to the `default_branch` property. That's odd ..

Finally I could track down that `resourceGitRepositoryUpdate` is called, regardless if you changed anything for the Git repository resource. `clients.GitReposClient.UpdateRepository` returns the error message.

**Comment by tmeckel** [comment]
@xuzhang3 Mistery solved. The value for the property `default_branch`must always be the complete Git path to the branch. So for branch `master` this is `refs/heads/master` and not simply `master`.  This also explains why Terraform wants to update the Gi repository resource. The REST API returns a different value for the `default_path` property than the HCL definition. So from my point of view we have two options here

1. Update the documentation of the Git Resource that the value for the `default_path` property has to always the complete Git path of the branch and not only it's name, or
2. Make code more resilient about the fact that users only specify the name of the branch. Like _if the value does not start with refs/heads prepend this to the value_

Presumably you don't wanna add me option 2. to this **monster** 👹 PR 😁 

Nevertheless I update the sample in the documentation around `azuredevops_git_permissions`

**Comment by xuzhang3** [comment]
@tmeckel  Thanks for your PR,  all feature is OK. 
For git permission management, I'd like to mention that if user set `Inheritance` opened, some configure may override by inherit permission settings.  This switch is opened by default. 
image

Example:
use the example configure in `git_permission`. I'm the admin, I cannot change the `CreateBranch` to `NotSet` If `Inheritance` is open, the `CreateBranch` can only be set to `allow`  or`deny`. This may get users confused.  Can you update the document that if `Inheritance` is open, permission configure may override by inherit.

**Comment by tmeckel** [comment]
> 
> 
> @tmeckel Thanks for your PR, all feature is OK.
> For git permission management, I'd like to mention that if user set `Inheritance` opened, some configure may override by inherit permission settings. This switch is opened by default.
> Example:
> use the example configure in `git_permission`. I'm the admin, I cannot change the `CreateBranch` to `NotSet` If `Inheritance` is open, the `CreateBranch` can only be set to `allow` or`deny`. This may get users confused. Can you update the document that if `Inheritance` is open, permission configure may override by inherit.

@xuzhang3 I think you refer to this behavior, right? I mean that if you reset a defined value to `NotSet` on a lower level, the permission will be set to the permission the principal gets assigned from a higher level in the web interface.

azdo-inherited-permissions

The value `NotSet` is explicitly intended to set the value of a permission to the parent value if inheritance is enabled, but you definitely define an ACE on that particular level (eg on a Git branch), even if the ACE is empty for a particular `action` (no values in `allow` and `deny`). I think a piece of code from the `namespaces.go` file makes this clear:

image

To emphasize it a bit more: every time you define a `azuredevops_git_permission` resource you'll create an ACE in the ACL for the security namespace token.

The fact that the web interface of Azure DevOps is nice enough to always show the user (admin) the effective permissions for an identity at one level may require explanation but is technically irrelevant for the provider.

Oh and not to forget, if you want to reset a `allow`or `deny` permission to the inherited permissions of a principal you **must** set the value to `NotSet` for an security namespace action in HCL.

And I don't want to be mean, but I can't find a single word about this whole topic in the official documentation of the REST API. So why should we make up for this deficit on our side?

**Comment by tmeckel** [comment]
@xuzhang3 @EliiseS @nmiodice  WOOOOHOOOOOO!!! Finally done!! 😃 👍 Thanks for never giving up, while reviewing this massive PR!

Celebrating

**Pull request merged**
